### PR TITLE
[Stan 2.33] Pathfinder

### LINKED ISSUE #684: [Stan 2.33] Pathfinder algorithm
## Summary:

Expose Pathfinder: https://github.com/stan-dev/stan/pull/3123 https://github.com/stan-dev/cmdstan/pull/1155

This issue is to sketch out a design and for coordinating with the CmdStanR team (@jgabry)

## Description:

Pathfinder is useful for standalone VI and for initializing MCMC, so we should cater to both usages.

#### New method on `CmdStanModel`: `pathfinder`.

Signature:

```python

    def pathfinder(
        self,
        data: Union[Mapping[str, Any], str, os.PathLike, None] = None,
        *,
        init_alpha: Optional[float] = None,
        tol_obj: Optional[float] = None,
        tol_rel_obj: Optional[float] = None,
        tol_grad: Optional[float] = None,
        tol_rel_grad: Optional[float] = None,
        tol_param: Optional[float] = None,
        history_size: Optional[int] = None,
        draws: Optional[int] = None, # controls size of actual output
        num_single_draws: Optional[int] = None, # corresponds to the num_draws argument on cmdline
        num_paths: Optional[int] = None,
        max_lbfgs_iters: Optional[int] = None,
        num_elbo_draws: Optional[int] = None,
       # arguments standard to all methods
        seed: Optional[int] = None,
        inits: Union[Dict[str, float], float, str, os.PathLike, None] = None,
        output_dir: OptionalPath = None,
        sig_figs: Optional[int] = None,
        save_profile: bool = False,
        show_console: bool = False,
        refresh: Optional[int] = None,
        time_fmt: str = "%Y%m%d%H%M%S",
        timeout: Optional[float] = None,
    ) -> CmdStanPathfinder:
```

This is probably more-or-less what you would expect going off of the CmdStan implementation. One notable thing is - at the moment - I am not providing a way to save individual pathfinder runs (the `save_single_paths` arg). If things from the diagnostic files produced by this argument end up being important to initializing MCMC, we can add support then, but I don't think it is valuable in the first implementation

#### New Class `CmdStanPathfinder`

This will look broadly similar to CmdStanLaplace, with the main methods being those that let you access the draws: `stan_variable`, `draws_pd`, etc

For supporting the initialization of MCMC, I think it is important to also have a method like 

`create_inits(self, seed = None, num_chains = 4)`

This will return a list of dictionaries of 4 random draws from the pathfinder outputs in the format required for the inits argument to other methods. We can also consider adding `CmdStanPathfinder` as an option for the `inits` argument directly, in which case this method would just be called.

### PR DESCRIPTION
**Pull request by WardBrian**
This PR is based on #681, which should be merged first. 

#### Submission Checklist

- [x] Run unit tests
- [x] Declare copyright holder and open-source license: see below

#### Summary

Closes #684.

- Adds a `pathfinder` method to CmdStanModel
- Adds argument validation for Pathfinder, cleans up other validation
- Add CmdStanPathfinder class with expected capabilities. Also improve CmdStanLaplace from #669

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):

Simons Foundation

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)

### CODE REVIEW

File: cmdstanpy/stanfit/pathfinder.py [hunk 1]
```diff
@@ -27,6 +27,40 @@ def __init__(self, runset: RunSet):
         config = scan_generic_csv(runset.csv_files[0])
         self._metadata = InferenceMetadata(config)
 
+    def create_inits(
+        self, seed: Optional[int] = None, chains: int = 4
+    ) -> Union[List[Dict[str, np.ndarray]], Dict[str, np.ndarray]]:
+        self._assemble_draws()
+        rng = np.random.default_rng(seed)
+        idxs = rng.choice(self._draws.shape[0], size=chains, replace=False)
+        if chains == 1:
```
[the remark this answers is not in the document]
Comment (WardBrian) [reply cmdstanpy/stanfit/pathfinder.py:36]: Yes. The `chains` argument exists to make it easy to initialize multiple chains of MCMC with different draws from the same pathfinder run

**Comment by WardBrian** [comment]
Note: https://github.com/stan-dev/cmdstanpy/issues/362 makes `create_inits` currently less easy to use, should be resolved

**Comment by WardBrian** [comment]
@mitzimorris @jgabry - I think this is essentially feature complete if we're happy with the interface from #684, mind starting to give it a glance?

File: cmdstanpy/cmdstan_args.py
Comment (mitzimorris) [inline cmdstanpy/cmdstan_args.py:1]: thanks for helper "positive_int"  - cleanup makes a real difference!

File: docsrc/api.rst [hunk 2]
```diff
@@ -25,11 +25,17 @@ A CmdStanModel object encapsulates the Stan program. It manages program compilat
...
 :meth:`~CmdStanModel.optimize`
-    produce a penalized maximum likelihood estimate (point estimate) of the model parameters.
+    produce a penalized maximum likelihood estimate or maximum a posteriori estimate (point estimate) of the model parameters.
+
+:meth:`~CmdStanModel.laplace_sample`
+    draw from a Laplace approximatation centered at the posterior mode found by ``optimize``.
 
 :meth:`~CmdStanModel.variational`
-    run CmdStan’s variational inference algorithm to approximate the posterior distribution.
+    run CmdStan’s ADVI variational inference algorithm to approximate the posterior distribution.
```
Comment (mitzimorris) [inline docsrc/api.rst:38]: spell out automatic-differentiation variational inference (ADVI) - or "ADVI algorithm" - prefer the former.

File: cmdstanpy/stanfit/pathfinder.py [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+        :param seed: Used for random selection, defaults to None
+        :param chains: Number of initial values to return, defaults to 4
+        :return: The initial values for the parameters of the model.
+            If ``chains`` is 1, a dictionary is returned, otherwise a list
+            of dictionaries is returned, in the format expected for
+            :meth:`CmdStanModel.sample`'s ``inits`` argument.
+        """
+        self._assemble_draws()
+        rng = np.random.default_rng(seed)
+        idxs = rng.choice(self._draws.shape[0], size=chains, replace=False)
```
Comment (mitzimorris) [inline cmdstanpy/stanfit/pathfinder.py:51]: yes! use of replace=False is correct.

**Review by mitzimorris — APPROVED** [review]
looks good!

**Comment by WardBrian** [comment]
@jgabry do you mind giving this a glance just to confirm you're happy with the interface/names of things so cmdstanr can match?

**Comment by WardBrian** [comment]
Jonah is busy this week so if @rok-cesnovar or @andrjohns could look at this and the design sketched in #684 for cmdstanr-parity reasons I'd appreciate it!

**Comment by mitzimorris** [comment]
hooray!!!!

**Pull request merged**